---
title: "Admixture"
author: "Enrico"
date: "28 September 2018"
output:
  pdf_document: default
  html_document: default
---

# 0. Path Definition

Define the paths to different locations for easier calling during script

```{r Define paths, eval=FALSE, engine='bash'}

ssh -X ebazzicalupo@genomics-b.ebd.csic.es # these analyses will be conducted in server B because the annotated VCF file is located there

Admx_PATH=/home/ebazzicalupo/Admixture # path to the project directory
Data_PATH=/home/ebazzicalupo/Data # path to data directory
ShapeIt=/home/ebazzicalupo/Admixture/ShapeIt/shapeit # path to ShapeIt software
Lenght_PATH=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/Length_scaffolds_lp23 # path to scaffold length file
SexChr_PATH=/home/GRUPOS/grupolince/sexual_chr # path to directory with lists of scaffolds with XY chromosomes represented
BCF=/opt/bcftools-1.6/bcftools # path to BCFtools software
REF=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23.fa # path to reference genome
GATK=/opt/GATK-3.7/GenomeAnalysisTK.jar # GATK software path
VCF2BEA=/home/ebazzicalupo/Admixture/VCF2Beagle/vcf2beagle.jar # vcf2beagle software path
PCAdmix=/home/ebazzicalupo/Admixture/PCAdmix/PCAdmix3_linux # PCAdmix software path

```

# Introgression Analysis with PCAdmix

# 1. Preparing all the data

## Selection of scaffolds to include in the analysis - REMEMBER to remove XY

PCAdmix runs on single chromosomes. Having scaffolds of various length, we decided to include only the bigger scaffolds with enough information to be relevant in the analysis. We decided on 1Mbp as the lower limit for the scaffold length. We decided not to include scaffolds that might contain parts of the sexual chromosomes, which might interefere with the analysis. For this reason we removed all of the scaffolds which are annotated for any part of their length to sexual chromosomes.

```{r Preparing all the data, eval=FALSE, engine='bash'}

# check which ones have more than 1Mbp
Lenght_PATH=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/Length_scaffolds_lp23
sort -nk2,2 $Lenght_PATH | tail -673 | sort -k1 > $Admx_PATH/lp23_scaffolds_longerthan1Mbp  # 673 have more than 1Mbp

# no XY chromosome
cut -f1 $SexChr_PATH/females_males_normalized_ratio.coverage.contig.filteredX.percontig.Xchr-above30.0based.bed | sort -u > $Admx_PATH/lp23_scaffolds_Xchromosome

cut -f1 $SexChr_PATH/females_males_normalized_ratio.coverage.contig.filteredY.lax.percontig.Ychr-above30.0based.bed | sort -u > $Admx_PATH/lp23_scaffolds_Ychromosome

cat $Admx_PATH/lp23_scaffolds_Xchromosome $Admx_PATH/lp23_scaffolds_Ychromosome | sort -u > $Admx_PATH/lp23_scaffolds_sexchromosome

# remove sexchromosome scaffolds from list of scaffolds longer than 1Mbp
grep -v -f $Admx_PATH/lp23_scaffolds_sexchromosome $Admx_PATH/lp23_scaffolds_longerthan1Mbp | cut -d ' ' -f1 > $Admx_PATH/lp23_scaffolds_tobephased

```

## Get VCF containing all common positions between Lynx lynx and Lynx pardinus (to keep all of the available information) and filter for coverage and missingness

We start with two VCF created by Dani, which are separated by species, but all of the positions which are fixed for the ancestral allel in one species and not in the other are not filtered out yet (Dani filters these positions for his analyses -> he tags it as "trimmed" in the name). We will apply the same coverage and missingness filters Dani applies to both VCFs, to filter out unwanted positions. To reduce computational times we filter to only keep the scaffolds we are interested in (see step above). We will divide these by the populations we are interested in (eastern (vl+ya+og+ka+to) and western (pv+cr+po+la+ki+ur) populations of Lynx lynx, and sierra morena population (sm) from Lynx pardinus). Once we have these VCFs, we will intersect them in order to keep only the common positions between the three. We can then go to the next step which is PHASING.

```{r Preparing all the data, eval=FALSE, engine='bash'}

# Copy the untrimmed VCF files of Lynx lynx (ll_perspecies.lr_ann.vcf) and of Lynx pardinus (lp_perspecies.lr_ann.vcf) from Dani's database (/home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/annotation/) to the data folder.

scp /home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/annotation/ll_perspecies.lr_ann.vcf $Data_PATH
scp /home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/annotation/lp_perspecies.lr_ann.vcf $Data_PATH

grep -v '#' $Data_PATH/ll_perspecies.lr_ann.vcf | wc -l #5063416
grep -v '#' $Data_PATH/lp_perspecies.lr_ann.vcf | wc -l #5063416

# Copy the depth calculations from Dani's database
scp /home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/c_lp_do-c_lp_sm_n031_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv $Data_PATH
scp /home/GRUPOS/grupolince/lynx_genomes_5x/VCFs_Dani/c_ll_no-c_ll_ba-h_ll_ba-c_ll_cr-c_ll_po-c_ll_la-c_ll_ki-c_ll_ur-c_ll_tu-c_ll_to-c_ll_ka-c_ll_og-c_ll_ya-c_ll_vl_n080_mean_sd_depthGlobal_lynx_per_pop_mean_folds_0.95.csv $Data_PATH

##Apply low and high DP and 0.325 missingness filters.
#First, for each species exclude those positions that have more than 32.5% missing genotypes (i.e. missing at least in all individuals within the 5x dataset, or within the smaller 25x dataset), as well as those that have lower (higher) depth than the minimum (maximum) within 0.95 of the distribution, as calculated above:

cd $Admx_PATH
screen -S perspeciesfilter.log
script perspeciesfilter.log

cd $Data_PATH
declare SPECIES=$(ls *_perspecies.lr_ann.vcf | cut -c1-2 | uniq)
for i in ${SPECIES[@]}
  do
  echo "${i}"
  MIN_DP=$(cat $Data_PATH/c_"${i}"*.csv | awk '{print $9}') #Obtained by Dani
  MAX_DP=$(cat $Data_PATH/c_"${i}"*.csv | awk '{print $8}') #Obtained by Dani
  echo $MIN_DP
  echo $MAX_DP
  $BCF filter -e "DP < ${MIN_DP} || DP > ${MAX_DP} || F_MISSING > 0.325" -Ov -o $Data_PATH/"${i}"_perspecies_filtered2.lr_ann.vcf $Data_PATH/"${i}"_perspecies.lr_ann.vcf
done
  
grep -v '#' $Data_PATH/lp_perspecies_filtered2.lr_ann.vcf | wc -l #4878018
grep -v '#' $Data_PATH/ll_perspecies_filtered2.lr_ann.vcf | wc -l #5010540

```

## Filter to be left only with "useful" scaffolds

Filter the VCFs to include only the previously selected scaffolds (longer than 1Mbp and not including X or Y annotations)

```{r Preparing all the data, eval=FALSE, engine='bash'}

cd $Data_PATH
declare SPECIES=$(ls *_perspecies.lr_ann.vcf | cut -c1-2 | uniq)
for i in ${SPECIES[@]}
  do
  grep '#' $Data_PATH/"${i}"_perspecies_filtered2.lr_ann.vcf > "${i}"_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf && grep -v '#' $Data_PATH/"${i}"_perspecies_filtered2.lr_ann.vcf | grep -f $Admx_PATH/lp23_scaffolds_tobephased >> "${i}"_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf
done

grep -v '#' $Data_PATH/lp_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | wc -l #2759315
grep -v '#' $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | wc -l #2786031

```

We will divide these VCFs by the populations we are interested in (eastern (vl+ya+og+ka+to) and western (pv+cr+po+la+ki+ur) populations of Lynx lynx, and sierra morena population (sm) from Lynx pardinus). These partitions were decided based on the PCA which separates largely between eastern and western populations of Lynx lynx. Some populations were not included in the groups because of characteristics which are not optimal for the analysis (Norway is more separated from the Carpathians than the rest, Balkans samples are not perfectly "clean", Tuva shows signs of admixture)
Once we have these VCFs, we will intersect them in order to keep only the common positions between the three. We can then go to the next step which is PHASING.

## Get Hybrid individual (pv = pais vasco) and WESTERN L.lynx populations' (cr = carpathians, po = poland, la = latvia, ki = kirov, ur = urali) VCF to phase

```{r Preparing all the data, eval=FALSE, engine='bash'}

## GATK to separate pv and cr individuals

cd $Admx_PATH
rm llwest_list_to_remove.txt

# define which individuals to extract
$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'pv\|cr\|po\|la\|ki\|ur' > $Admx_PATH/llwest_list_to_remove.txt
cat $Admx_PATH/llwest_list_to_remove.txt

# GATK
screen -S GATKpartitionWEST.log
script GATKpartitionWEST.log

java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -Xms16g -Xmx32g -jar $GATK \
    -T SelectVariants \
    -R $REF \
    -V $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf \
    -o $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf \
    --sample_file $Admx_PATH/llwest_list_to_remove.txt

rm $Admx_PATH/llwest_list_to_remove.txt

grep -v "#" $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | wc -l #2786031

```

## Get EASTERN Lynx lynx population (vl = vladivostok, ya = yakutia, og+ka+to = mongolia) VCF to phase

```{r Preparing all the data, eval=FALSE, engine='bash'}

## GATK to separate vl individuals from lynx lynx perspecies vcf

cd $Admx_PATH
rm lleast_list_to_remove.txt

# define which individuals to extract
$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'vl\|ya\|og\|ka\|to' > $Admx_PATH/lleast_list_to_remove.txt
cat $Admx_PATH/lleast_list_to_remove.txt

# GATK
screen -S GATKpartitionEAST.log
script GATKpartitionEAST.log

java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -Xms16g -Xmx32g -jar $GATK \
    -T SelectVariants \
    -R $REF \
    -V $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf \
    -o $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf \
    --sample_file $Admx_PATH/lleast_list_to_remove.txt

rm $Admx_PATH/lleast_list_to_remove.txt

grep -v "#" $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | wc -l #2786031

```

## Get Sierra Morena VCF to phase

```{r Preparing all the data, eval=FALSE, engine='bash'}

## GATK to separate sm individuals from lynx lynx perspecies vcf

cd $Admx_PATH
rm sm_list_to_remove.txt

# define which individuals to extract
$BCF query -l $Data_PATH/lp_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'sm' > $Admx_PATH/sm_list_to_remove.txt
cat $Admx_PATH/sm_list_to_remove.txt

# GATK
screen -S GATKpartitionSM.log
script GATKpartitionSM.log

java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -Xms16g -Xmx32g -jar $GATK \
    -T SelectVariants \
    -R $REF \
    -V $Data_PATH/lp_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf \
    -o $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf \
    --sample_file $Admx_PATH/sm_list_to_remove.txt

rm $Admx_PATH/sm_list_to_remove.txt

grep -v "#" $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | wc -l #2759315

```

## Intersect the three VCFs

To have each VCF with only the positions which are common for the three. Both BCFtools and BEDtools have been tried. Since same species have same positions we can simply intersect with BEDtools (easier) the files 2 by 2 comparing sm with one of the two ll populations and the two ll populations with sm.

```{r Preparing all the data, eval=FALSE, engine='bash'}

cd $Data_PATH
# index bgziped VCFs
$BCF index lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz
$BCF index ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz
$BCF index ll_west_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz

## intersect with BCFtools results in separate folder
$BCF isec -p $Data_PATH/intersect -n=3 lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz ll_west_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz

#This file was produced by vcfisec.
#The command line was:   bcftools isec  -p /home/ebazzicalupo/Data/intersect -n=3 lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz ll_vl_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz ll_pv.cr_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz

#Using the following file names:
# /home/ebazzicalupo/Data/intersect/0000.vcf      for stripped    lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz
# /home/ebazzicalupo/Data/intersect/0001.vcf      for stripped    ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz
# /home/ebazzicalupo/Data/intersect/0002.vcf      for stripped    ll_west_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf.gz

# move results to Datapath
cd $Data_PATH/intersect
mv 0000.vcf $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds_intersect.lr_ann.vcf
mv 0001.vcf $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds_intersect.lr_ann.vcf
mv 0002.vcf $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds_intersect.lr_ann.vcf

# check number of SNPs
grep -v "#" $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds_intersect.lr_ann.vcf | wc -l #2752014
grep -v "#" $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds_intersect.lr_ann.vcf | wc -l #2752014
grep -v "#" $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds_intersect.lr_ann.vcf | wc -l #2752014

## Intersect with BEDtools - Used BEDtools in the end
# sm population
bedtools intersect -header -a $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf -b $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf > $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf

# east population
bedtools intersect -header -a $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf -b $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf > $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf

# west population
bedtools intersect -header -a $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf -b $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds.lr_ann.vcf > $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf

# check number of SNPs
grep -v "#" $Data_PATH/lp_sm_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | wc -l #2752014
grep -v "#" $Data_PATH/ll_east_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | wc -l #2752014
grep -v "#" $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | wc -l #2752014

```

# 2. PHASING

## trials

```{r Phasing, eval=FALSE, engine='bash'}

## The final filtered VCF file of the doñana population (c_lp_do_perpop_with1.lr_ann.vcf) was copied from Dani's database (/VCFs_Dani/annotation/c_lp_do_perpop folder) to the project folder.

cd $Admx_PATH

# create trial file
grep "#" $Data_PATH/ll_vl_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf > $Admx_PATH/Phasing/trial.vcf && grep -v "#" $Data_PATH/ll_vl_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | grep "lp23.s00001" >> $Admx_PATH/Phasing/trial.vcf

$ShapeIt -V trial.vcf \
        --window 0.5 \
        --output-graph trial.graph \
        --force \
        --states 200
        
$ShapeIt -convert \
        --input-graph trial.graph \
        --output-max trial.phased.haps trial.phased.sample
        
$ShapeIt -convert \
        --input-haps trial.phased \
        --output-vcf trial.phased.vcf

# count number of scaffolds
grep -v "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf | cut -f1 | sort -u | wc -l
grep -v "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf | cut -f1 | sort -u | head -1

# count number of SNPs in scaffold
awk '$1 == "lp23.s41442" { count++ } END { print count }' $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf

```

## loop trial - explanation in trial

```{r Phasing, eval=FALSE, engine='bash'}

# for doñana only${scaffold}

cd $Admx_PATH/Phasing/

all_scaffolds=$(grep -v "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf | cut -f1 | sort -u)
for scaffold in $all_scaffolds
  do
  if [ $(grep -v "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf | grep "${scaffold}" | wc -l) -ge 12 ] ;
  then
    echo "Starting Phasing of scaffold ${scaffold}"
    
    mkdir $Admx_PATH/Phasing/${scaffold}_Phase
    
    rm $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.vcf
   
    ##ERROR grep -v "#" should be done before grepping scaffold
    grep "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf > $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.vcf && grep "${scaffold}" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf >> $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.vcf 
    
    $ShapeIt -V $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.vcf \
        --window 0.5 \
        --output-graph $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.graph \
         --force \
        --states 200
        
    $ShapeIt -convert \
        --input-graph $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.graph \
        --output-max $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.phased

    $ShapeIt -convert \
        --input-haps $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.phased \
        --output-vcf $Admx_PATH/Phasing/${scaffold}_Phase/${scaffold}.phased.vcf
  fi
done

## REMOVE RESULTS OF 1 SNP (Because previous code was run wrong -> no if statement)

all_scaffolds=$(grep -v "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf | cut -f1 | sort -u)
for scaffold in $all_scaffolds
  do
  if [ $(grep -v "#" $Data_PATH/c_lp_do_perpop_with1.lr_ann.vcf | grep "${scaffold}" | wc -l) -eq 1 ] ;
    then
    rm -r $Admx_PATH/Phasing/${scaffold}_Phase
  fi
done

## CONCATENATE ALL PHASED VCFs into one file
cd $Admx_PATH/Phasing/

phased_scaffolds=$(ll -htr | grep -o "lp23.*_Phase" | sed 's/_Phase//g')
rm c_lp_do_perpop_phased.vcf
grep "#" $Admx_PATH/Phasing/lp23.s00001_Phase/lp23.s00001.phased.vcf > $Admx_PATH/Phasing/c_lp_do_perpop_phased.vcf

for scaffold in $phased_scaffolds
  do ....

```

## Loop for sm (sierra morena), vl (vladivostok), and pv+cr (pais vasco + carpathians) - explanation in trial

```{r Phasing, eval=FALSE, engine='bash'}

# As Server B was full all the data was copied to server A and this loop was run there and everything from here was run on server A - the whole analysis took 13 hours

cd $Admx_PATH
mkdir Phasing

screen -S PerPOPphasing.log
script PerPOPphasing.log

cd $Data_PATH/
declare POP=$(ls *_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | cut -d'_' -f1,2 | uniq)
echo $POP
cd $Admx_PATH/Phasing

for pop in $POP
  do
  echo "Starting Phasing of population ${pop}"
  scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased)
  for scaffold in $scaffolds
    do
    echo "Starting Phasing of scaffold ${scaffold} in population ${pop}"
    
    mkdir $Admx_PATH/Phasing/${pop}_${scaffold}_Phase
    
    grep "#" $Data_PATH/${pop}_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf > $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.vcf && grep -v "#" $Data_PATH/${pop}_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | grep "${scaffold}" >> $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.vcf
    
    $ShapeIt -V $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.vcf \
        --window 0.5 \
        --output-graph $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.graph \
        --force \
        --states 500
        
    $ShapeIt -convert \
        --input-graph $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.graph \
        --output-max $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.phased

    $ShapeIt -convert \
        --input-haps $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.phased \
        --output-vcf $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.phased.vcf
  done
done

## There was a problem with phasing of ll_west_s36513 so I tried phasing individually - STILL DIDN'T WORK
## error message : shapeit: src/modes/phaser/phaser_algorithm.cpp:150: void phaser::phaseSegment(int): Assertion `conditional_index[segment].size() >= 2' failed. Abortado
$ShapeIt -V $Admx_PATH/Phasing/ll_west_lp23.s36513_Phase/ll_west_lp23.s36513.vcf \
        --window 0.5 \
        --output-graph $Admx_PATH/Phasing/ll_west_lp23.s36513_Phase/ll_west_lp23.s36513.graph \
        --force \
        --thread 1 \
        --states 500
        

## Remove s36513 and create new scaffolds file without it

cd $Data_PATH/
declare POP=$(ls *_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf | cut -d'_' -f1,2 | uniq)
echo $POP
cd $Admx_PATH/Phasing/
for pop in $POP
  do
  echo "Removing lp23.s36513 from ${pop}"
  rm -r $Admx_PATH/Phasing/${pop}_lp23.s36513_Phase
  echo " DONE "
done

grep -v "lp23.s36513" $Admx_PATH/lp23_scaffolds_tobephased > $Admx_PATH/lp23_scaffolds_tobephased_noerror

```

## Separate pv individual from west population's phased VCFs

```{r Phasing, eval=FALSE, engine='bash'}

## try with one
bgzip $Admx_PATH/Phasing/ll_west_lp23.s00001_Phase/ll_west_lp23.s00001.phased.vcf

tabix -p vcf $Admx_PATH/Phasing/ll_west_lp23.s00001_Phase/ll_west_lp23.s00001.phased.vcf.gz

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'pv' > $Admx_PATH/pv_list_to_remove.txt
cat $Admx_PATH/pv_list_to_remove.txt

$BCF view -Ov -S $Admx_PATH/pv_list_to_remove.txt $Admx_PATH/Phasing/ll_west_lp23.s00001_Phase/ll_west_lp23.s00001.phased.vcf.gz > $Admx_PATH/Phasing/ll_west_lp23.s00001_Phase/ll_pv_lp23.s00001.phased.vcf

## loop for all

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'pv' > $Admx_PATH/pv_list_to_remove.txt
cat $Admx_PATH/pv_list_to_remove.txt

cd $Admx_PATH
screen -S PVpop.log
script PVpop.log

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo ${scaffold}
    
    bgzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf
    tabix -p vcf $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz

    rm -r $Admx_PATH/Phasing/ll_pv_${scaffold}_Phase
    mkdir $Admx_PATH/Phasing/ll_pv_${scaffold}_Phase
    
    $BCF view -Ov -S $Admx_PATH/pv_list_to_remove.txt $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz > $Admx_PATH/Phasing/ll_pv_${scaffold}_Phase/ll_pv_${scaffold}.phased.vcf
    gunzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    echo " DONE "

done

```

## Beagle format conversion for PCAdmix

```{r Phasing, eval=FALSE, engine='bash'}

# the program vcf2beagle.jar was copied in Admixture/VCF2Beagle folder from my laptop's Desktop
scp ~/Desktop/vcf2beagle.jar ebazzicalupo@genomics-a.ebd.csic.es:/home/ebazzicalupo/Admixture/VCF2Beagle

# try for one

cat $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased.vcf  | java -jar $VCF2BEA N $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased

# loop for all
screen -S vcf2beagle.log
script vcf2beagle.log

cd $Admx_PATH/Phasing
declare POP=$(ls *_Phase | cut -d '_' -f1,2 | sort -u)
echo $POP

for pop in $POP
  do
  scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
  for scaffold in $scaffolds
    do
    echo  " converting to beagle ${scaffold} in ${pop} "
    cat $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.phased.vcf | java -jar $VCF2BEA N $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.phased
    gunzip $Admx_PATH/Phasing/${pop}_${scaffold}_Phase/${pop}_${scaffold}.phased.bgl.gz
    echo " DONE "
  done
done
    
```

# 3. PCAdmix

```{r PCAdmix, eval=FALSE, engine='bash'}

# try with one scaffold
mkdir $Admx_PATH/PCAdmix_results/

$PCAdmix -anc $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased.bgl $Admx_PATH/Phasing/lp_sm_lp23.s00001_Phase/lp_sm_lp23.s00001.phased.bgl -adm $Admx_PATH/Phasing/ll_pv_lp23.s00001_Phase/ll_pv_lp23.s00001.phased.bgl -o $Admx_PATH/PCAdmix_results/pcadmix_lp23.s00001 -lab L.pardinus L.lynx Admixed -wMb 0.1

# loop for all scaffolds
# WITHOUT PUTTING A MAP IT WILL USE DEFAULT WINDOW SIZE OF 20 SNPs (-wMb 0.1 is useless without a map)

screen -S PCAdmix.log
script PCAdmix.log

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_results/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_pv_${scaffold}_Phase/ll_pv_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_results/PCAdmix_${scaffold}/pcadmix_${scaffold} -lab LYNX PARDINUS Admixed -wMb 0.1
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_results/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_results/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_results/PCAdmix_lp23.s*/*vit.txt > ~/home/allscaffolds.txt

grep "1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1" ~/home/allscaffolds.txt | wc -l

```

## Version with physical distance map

```{r PCAdmix, eval=FALSE, engine='bash'}

## We first need to creat a physical map in PLINK format (chromosome number, SNP ID, genetic position [optional], phsyical position)

## try for one scaffold

# Generate MAP

paste -d"\t" <(grep -v "#" $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased.vcf | cut -f1) <(grep -v "id" $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased.bgl | cut -d' ' -f2) <(grep -v "#" $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased.vcf | cut -f2) > $Admx_PATH/maptry.map

while read a b c
do
echo -e $a '\t' $b '\t' 0 '\t' $c
done <$Admx_PATH/maptry.map > tmp && mv tmp $Admx_PATH/maptry.map

# run PCAdmix

$PCAdmix -anc $Admx_PATH/Phasing/ll_east_lp23.s00001_Phase/ll_east_lp23.s00001.phased.bgl $Admx_PATH/Phasing/lp_sm_lp23.s00001_Phase/lp_sm_lp23.s00001.phased.bgl -adm $Admx_PATH/Phasing/ll_pv_lp23.s00001_Phase/ll_pv_lp23.s00001.phased.bgl -map $Admx_PATH/maptry_with0.map -lab L.pardinus L.lynx Admixed -o $Admx_PATH/pcadmix_withmap_lp23.s00001 -wMb 0.1

## LOOP for all scaffolds

mkdir $Admx_PATH/ScaffoldMaps
mkdir $Admx_PATH/PCAdmix_hundredKb/

screen -S PCAdmix_hundredKb.log
script PCAdmix_hundredKb.log

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_hundredKb/PCAdmix_${scaffold}
    
    echo " creating ${scaffold} map "
    
    paste -d"\t" <(grep -v "#" $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.vcf | cut -f1) <(grep -v "id" $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl | cut -d' ' -f2) <(grep -v "#" $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.vcf | cut -f2) > $Admx_PATH/ScaffoldMaps/${scaffold}.map
    while read a b c
      do
      echo -e $a '\t' $b '\t' 0 '\t' $c
    done <$Admx_PATH/ScaffoldMaps/${scaffold}.map > tmp && mv tmp $Admx_PATH/ScaffoldMaps/${scaffold}.map
    
    echo " running PCAdmix "
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_pv_${scaffold}_Phase/ll_pv_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_hundredKb/PCAdmix_${scaffold}/pcadmix_${scaffold} -map $Admx_PATH/ScaffoldMaps/${scaffold}.map -lab LYNX PARDINUS Admixed -wMb 0.1
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_hundredKb/PCAdmix_lp23.s*/*vit.txt | less -S
# save in new file on laptop
cat ~/home/PCAdmix_hundredKb/PCAdmix_lp23.s*/*vit.txt > ~/home/hundredKb.txt
cat $Admx_PATH/PCAdmix_hundredKb/PCAdmix_lp23.s*/*ia.txt | less -S


```

## Version with Carpathians population as Admixed (for results comparison)

```{r PCAdmix, eval=FALSE, engine='bash'}

# extract carpathians (cr) population vcf

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'cr' > $Admx_PATH/cr_list_to_remove.txt
cat $Admx_PATH/cr_list_to_remove.txt

cd $Admx_PATH
screen -S CRpop.log
script CRpop.log

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo ${scaffold}
    rm $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz.tbi
    bgzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf
    tabix -p vcf $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    rm -r $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase
    mkdir $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase
    
    $BCF view -Ov -S $Admx_PATH/cr_list_to_remove.txt $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz > $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.vcf
    gunzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    echo " DONE "

done

## convert to beagle

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo  " converting to beagle ${scaffold} in cr "
    cat $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.vcf | java -jar $VCF2BEA N $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased
    gunzip $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.bgl.gz
    echo " DONE "
done

## run 20SNPs window PCAdmix

screen -S CR_PCAdmix.log
script CR_PCAdmix.log
mkdir $Admx_PATH/PCAdmix_Carpathians

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_Carpathians/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_Carpathians/PCAdmix_${scaffold}/pcadmix_${scaffold} -lab LYNX PARDINUS Admixed -w 20
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_Carpathians/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_Carpathians/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_Carpathians/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/Carpathians.txt

```

## Version with Kirov population as Admixed (for results comparison)

```{r PCAdmix, eval=FALSE, engine='bash'}
# extract kirov (ki) population vcf

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'ki' > $Admx_PATH/ki_list_to_remove.txt
cat $Admx_PATH/ki_list_to_remove.txt

cd $Admx_PATH
screen -S KIpop.log
script KIpop.log

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo ${scaffold}
    rm $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz.tbi
    bgzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf
    tabix -p vcf $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    rm -r $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase
    mkdir $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase
    
    $BCF view -Ov -S $Admx_PATH/ki_list_to_remove.txt $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz > $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.vcf
    gunzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    echo " DONE "

done

## convert to beagle

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo  " converting to beagle ${scaffold} in ki "
    cat $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.vcf | java -jar $VCF2BEA N $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased
    gunzip $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.bgl.gz
    echo " DONE "
done

## run 20SNPs window PCAdmix

screen -S ki_PCAdmix.log
script ki_PCAdmix.log
mkdir $Admx_PATH/PCAdmix_Kirov

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_Kirov/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_Kirov/PCAdmix_${scaffold}/pcadmix_${scaffold} -lab LYNX PARDINUS Admixed -w 20
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_Kirov/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_Kirov/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_Kirov/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/Kirov.txt

```

## Version with CR+PV as the same Admixed population

```{r PCAdmix, eval=FALSE, engine='bash'}

# extract carpathians (cr) and pv populations vcf

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'cr\|pv' > $Admx_PATH/pvcr_list_to_remove.txt
cat $Admx_PATH/pvcr_list_to_remove.txt

cd $Admx_PATH
screen -S PVCRpop.log
script PVCRpop.logcd

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo ${scaffold}
    rm $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz.tbi
    bgzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf
    tabix -p vcf $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    rm -r $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase
    mkdir $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase
    
    $BCF view -Ov -S $Admx_PATH/pvcr_list_to_remove.txt $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz > $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased.vcf
    gunzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    echo " DONE "

done

## convert to beagle

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo  " converting to beagle ${scaffold} in pvcr "
    cat $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased.vcf | java -jar $VCF2BEA N $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased
    gunzip $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased.bgl.gz
    echo " DONE "
done

## run 50Kbp window PCAdmix

screen -S PVCR_PCAdmix.log
script PVCR_PCAdmix.log
rm -r $Admx_PATH/PCAdmix_PVCRresults
mkdir $Admx_PATH/PCAdmix_PVCRresults

scaffolds=$(grep -v "lp23.s00036" $Admx_PATH/lp23_scaffolds_tobephased_noerror) # was stopping at s00036
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    rm -r $Admx_PATH/PCAdmix_PVCRresults/PCAdmix_${scaffold}
    mkdir $Admx_PATH/PCAdmix_PVCRresults/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_PVCRresults/PCAdmix_${scaffold}/pcadmix_${scaffold} -map $Admx_PATH/ScaffoldMaps/${scaffold}.map -lab LYNX PARDINUS Admixed -wMb 0.05
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_PVCRresults/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_PVCRresults/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_PVCRresults/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/PVCR.txt


## run 50 SNPs window PCAdmix

screen -S PVCR50snps_PCAdmix.log
script PVCR50snps_PCAdmix.log
rm -r $Admx_PATH/PCAdmix_PVCR50snpsresults
mkdir $Admx_PATH/PCAdmix_PVCR50snpsresults

scaffolds=$(grep -v "lp23.s00036" $Admx_PATH/lp23_scaffolds_tobephased_noerror) # lp23.s00036 blocking the analysis
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    rm -r $Admx_PATH/PCAdmix_PVCR50snpsresults/PCAdmix_${scaffold}
    mkdir $Admx_PATH/PCAdmix_PVCR50snpsresults/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_PVCR50snpsresults/PCAdmix_${scaffold}/pcadmix_${scaffold} -map $Admx_PATH/ScaffoldMaps/${scaffold}.map -lab LYNX PARDINUS Admixed -w 50
    
    echo " DONE "
done

# MAP FILES ARE ONLY UP TO s36533!!

cat $Admx_PATH/PCAdmix_PVCR50snpsresults/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_PVCR50snpsresults/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_PVCR50snpsresults/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/PVCR50snps.txt

cat PCAdmix_lp23.s00048/pcadmix_lp23.s00048.markers.txt | cut -f1 | tail -1


## run 20 SNPs window PCAdmix

screen -S PVCR20snps_PCAdmix.log
script PVCR20snps_PCAdmix.log
rm -r $Admx_PATH/PCAdmix_PVCR20snpsresults
mkdir $Admx_PATH/PCAdmix_PVCR20snpsresults

scaffolds=$(grep -v "lp23.s00036" $Admx_PATH/lp23_scaffolds_tobephased_noerror) # lp23.s00036 blocking the analysis
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    rm -r $Admx_PATH/PCAdmix_PVCR20snpsresults/PCAdmix_${scaffold}
    mkdir $Admx_PATH/PCAdmix_PVCR20snpsresults/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_pvcr_${scaffold}_Phase/ll_pvcr_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_PVCR20snpsresults/PCAdmix_${scaffold}/pcadmix_${scaffold} -map $Admx_PATH/ScaffoldMaps/${scaffold}.map -lab LYNX PARDINUS Admixed -w 20
    
    echo " DONE "
done

# MAP FILES ARE ONLY UP TO s36533!!

cat $Admx_PATH/PCAdmix_PVCR20snpsresults/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_PVCR20snpsresults/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_PVCR20snpsresults/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/PVCR20snps.txt

cat ~/home/PCAdmix_PVCR20snpsresults/PCAdmix_lp23.s*/*vit.txt > ~/home/PVCR20snps.txt

```

# 4. Tables

Choosed 20SNPs window (better resolution and amount of information when compared to physical size windows). Separate populations better, as running PCAdmix with PV+CR introduced noise (most probably due to LD calculations made by the program). Table 1: scaffold number, Scaffold length (from first SNP in first window to last SNP in last window in markers.txt file), number of windows (not including windows smaller than 20 SNPs in markers.txt file), number of lp windows (in vit.txt), % of introgression at scaffold level (in ia.txt file), % introgression at window level (dividing totwindows by lpwindows). Window level Table 2: Scaffold?, Window number (not including windows smaller than 20 SNPs in markers.txt file), Ancestry of A allele, Ancestry of B allele , Start SNP position, End SNP position. Introgressed segments table (Table 3): scaffold number, allele, number of consecutive windows in segment, which windows, first window first SNP (start), last window last SNP (end), segment length (end - start).

## Table 1

```{r Tables, eval=FALSE, engine='bash'}

# Remove Windows of less than 20 SNPs
awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | less -S
rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | less -S # 3 because the file has space at the end of each line

# scaffold number

# Calculate scaffold length - 1st SNP in first window and last SNP in last window
awk -v FS=" " 'NF>20' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f1 | head -1 | cut -d ":" -f2 # 6927 - scaffold START
awk -v FS=" " 'NF>20' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f20 | tail -1 | cut -d ":" -f2 # 13182611 - scaffold END

# LENGTH
START=$(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f1 | head -1 | cut -d ":" -f2 | bc) # 6927 - scaffold START
END=$(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f20 | tail -1 | cut -d ":" -f2 | bc) # 13182611 - scaffold END

echo "$END-$START" | bc


# Total number of windows
awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | wc -l # 689

# Number of introgressed windows (Lynx pardinus ancestry)
rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_A | tr ' ' '\n' | grep -v "h_ll_pv_0223_A" | grep "1" | wc -l # 19 = number of 1 (lp) ancestries for haplotype A of h_ll_pv_0223
rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_B | tr ' ' '\n' | grep -v "h_ll_pv_0223_B" | grep "1" | wc -l # 19 = number of 1 (lp) ancestries for haplotype B of h_ll_pv_0223

# % of introgression at scaffold level (as calculated by the PCAdmix program)
cat pcadmix_lp23.s00001.ia.txt | grep h_ll_pv_0223_A | cut -d " " -f3 # 0.0748933
cat pcadmix_lp23.s00001.ia.txt | grep h_ll_pv_0223_B | cut -d " " -f3 # 0.0767638

# % of introgression at window level (as calculated by dividing the number of introgressed windows by the total number of windows)

TOTW=$(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | wc -l | bc) # 689

ALPW=$(rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_A | tr ' ' '\n' | grep -v "h_ll_pv_0223_A" | grep "1" | wc -l | bc) # 19 = number of 1 (lp) ancestries for haplotype A of h_ll_pv_0223

echo "scale=4; $ALPW/$TOTW" | bc | sed 's/^\./0./'

# errors try


```

## Table 2

```{r Tables, eval=FALSE, engine='bash'}

# scaffold number

# window number
paste -d '.' <(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | cut -f1 | grep -o "Window") <(seq -f "%03g" 1 $(awk -v FS=" " 'NF>20' pcadmix_lp23.s00001.markers.txt | wc -l))

# Ancestry of A allele
rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_A | tr ' ' '\n' | grep -v "h_ll_pv_0223_A"

# Ancestry of B allele
rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_B | tr ' ' '\n' | grep -v "h_ll_pv_0223_B"

# Start position
awk -v FS=" " 'NF>20' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f1 | cut -d ":" -f2

# End position 
awk -v FS=" " 'NF>20' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f20 | cut -d ":" -f2

paste <(paste -d '.' <(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | cut -f1 | grep -o "Window") <(seq -f "%03g" 1 $(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | wc -l))) <(rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_A | tr ' ' '\n' | grep -v "h_ll_pv_0223_A") <(rev pcadmix_lp23.s00001.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_B | tr ' ' '\n' | grep -v "h_ll_pv_0223_B") <(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f1 | cut -d ":" -f2) <(awk -v FS=" " 'NF>19' pcadmix_lp23.s00001.markers.txt | cut -f2- | cut -d " " -f20 | cut -d ":" -f2) > table2.txt

# s05219 (has 3 segments of 2 windows)
paste <(paste -d '.' <(awk -v FS=" " 'NF>20' pcadmix_lp23.s05219.markers.txt | cut -f1 | grep -o "Window") <(seq -f "%03g" 1 $(awk -v FS=" " 'NF>20' pcadmix_lp23.s05219.markers.txt | wc -l))) <(rev pcadmix_lp23.s05219.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_A | tr ' ' '\n' | grep -v "h_ll_pv_0223_A") <(rev pcadmix_lp23.s05219.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_B | tr ' ' '\n' | grep -v "h_ll_pv_0223_B") <(awk -v FS=" " 'NF>20' pcadmix_lp23.s05219.markers.txt | cut -f2- | cut -d " " -f1 | cut -d ":" -f2) <(awk -v FS=" " 'NF>20' pcadmix_lp23.s05219.markers.txt | cut -f2- | cut -d " " -f20 | cut -d ":" -f2) > table2.txt

```

## Table 3

```{r Tables, eval=FALSE, engine='bash'}

# scaffold number

# first cut 1,2 -> A allele; 1,3 -> B allele
cut -f1,2 table2.txt | tr '\t' '_' | tr '\n' ' ' | grep -o -e "Window...._0 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._0" | cut -d ' ' -f2- | rev | cut -d ' ' -f2- | rev | sed -e 's/_1//g' > A_9

cut -f1,3 table2.txt | tr '\t' '_' | tr '\n' ' ' | grep -o -e "Window...._0 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._1 Window...._0" | cut -d ' ' -f2- | rev | cut -d ' ' -f2- | rev | sed -e 's/_1//g' > B_9

# NEED FILE (starts and ends) with multiple same sized segments (s05219 (has 3 segments of 2 windows)) -> all will be done with file to be sure 

cut -f1,2 table2.txt | tr '\t' '_' | tr '\n' ' ' | grep -o -e "Window...._0 Window...._1 Window...._1 Window...._0" | cut -d ' ' -f2- | rev | cut -d ' ' -f2- | rev | sed -e 's/_1//g' > A_2

cut -d ' ' -f1 A_2 > A_2_starts
grep -f A_2_starts table2.txt | cut -f4

rev A_2 | cut -d ' ' -f1 | rev > A_2_ends
grep -f A_2_ends table2.txt | cut -f5

# Add Segment length to Window file
cat A_2 | tr ' ' '-' | awk -vFS="-" '{print NF, $0}' | tr ' ' '\t' > A_2_windows

# SHOULD BE DONE AS A LOOP FOR SEGMENTS OF SIZE FROM 1 to MAX (total number of windows in that scaffold)

nwin=$(rev pcadmix_lp23.s05219.vit.txt | cut -d " " -f3- | rev | grep h_ll_pv_0223_A | tr ' ' '\n' | grep -v "h_ll_pv_0223_A" | grep "1" | wc -l)
for n in $(seq 1 $nwin)
  do
  var=$(yes "Window...._1" | head -$n)
  lvar=$(echo $var)
  cut -f1,2 table2.txt | tr '\t' '_' | tr '\n' ' ' | grep -o -e "Window...._0 $lvar Window...._0" | cut -d ' ' -f2- | rev | cut -d ' ' -f2- | rev | sed -e 's/_1//g' > A_${n}
done

# paste everything - SHOULD ADD SCAFFOLD number AND ALLELE (A or B) -> done with yes
paste A_2_windows <(grep -f A_2_starts table2.txt | cut -f4) <(grep -f A_2_ends table2.txt | cut -f5) > A_2_table3.txt

# order by columns 1, 2 and 5 the final table (scaffold, allele and start position, in order to have segments in physical order)

sort -k2,2 -k1,1 -k5,5n Admixture/PCAdmix_results/h_ll_pv_0223_Table3.txt > tmp && mv tmp Admixture/PCAdmix_results/h_ll_pv_0223_Table3.txt

# calculate segment length (END - START)

awk 'BEGIN { OFS = "\t" } { $7 = $6 - $5 } 1' <Admixture/PCAdmix_results/h_ll_pv_0223_Table3.txt > tmp && mv tmp Admixture/PCAdmix_results/h_ll_pv_0223_Table3.txt

```

## Three Tables generator executable Script

This is the script of the executable bash program that will run generate the tables of results. While executing the program two variables should be fed into the program : $1 = Name of the directory of the PCAdmix results; $2 = Name (ID) of the individual of interest.

```{r Tables, eval=FALSE, engine='bash'}
#!/bin/bash

Admx_PATH=/home/ebazzicalupo/Admixture # path to the project directory
Data_PATH=/home/ebazzicalupo/Data # path to data directory
ShapeIt=/home/ebazzicalupo/Admixture/ShapeIt/shapeit # path to ShapeIt software
Lenght_PATH=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/Length_scaffolds_lp23 # path to scaffold length file
SexChr_PATH=/home/GRUPOS/grupolince/sexual_chr # path to directory with lists of scaffolds with XY chromosomes represented
BCF=/opt/bcftools-1.6/bcftools # path to BCFtools software
REF=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23.fa # path to reference genome
GATK=/opt/GATK-3.7/GenomeAnalysisTK.jar # GATK software path
VCF2BEA=/home/ebazzicalupo/Admixture/VCF2Beagle/vcf2beagle.jar # vcf2beagle software path
PCAdmix=/home/ebazzicalupo/Admixture/PCAdmix/PCAdmix3_linux # PCAdmix software path

## Table 1

cd $Admx_PATH/$1

rm $2_*.txt

scaffolds=$(ls -l | grep -o "PCAdmix_lp23.s....." | sed 's/PCAdmix_//g')

echo " - Generating Table 1 for $2 in $1... - "

echo -e "INDIVIDUAL\tSCAFFOLD\tSTART\tEND\tLENGTH\tTOT_WIN\tA_LP_WIN\tB_LP_WIN\tA_PCA_INTRO\tB_PCA_INTRO\tA_WPROP_INTRO\tB_WPROP_INTRO" > $2_Table1.txt

for scaffold in $scaffolds
    do
    
    echo -ne " - ${scaffold} - \r"
    
    INDIVIDUAL=$(echo "$2")
    SCAFFOLD=$(echo "${scaffold}")
    START=$(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | cut -f2- | cut -d " " -f1 | head -1 | cut -d ":" -f2 | bc)
	END=$(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | cut -f2- | cut -d " " -f20 | tail -1 | cut -d ":" -f2 | bc)
    LENGTH=$(echo "$END-$START" | bc)
    TOT_WIN=$(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | wc -l)
    A_LP_WIN=$(rev PCAdmix_${scaffold}/pcadmix_${scaffold}.vit.txt | cut -d " " -f3- | rev | grep $2_A | tr ' ' '\n' | grep -v "$2_A" | grep "1" | wc -l | bc)
    B_LP_WIN=$(rev PCAdmix_${scaffold}/pcadmix_${scaffold}.vit.txt | cut -d " " -f3- | rev | grep $2_B | tr ' ' '\n' | grep -v "$2_B" | grep "1" | wc -l | bc)
    A_PCA_INTRO=$(cat PCAdmix_${scaffold}/pcadmix_${scaffold}.ia.txt | grep $2_A | cut -d " " -f3)
    B_PCA_INTRO=$(cat PCAdmix_${scaffold}/pcadmix_${scaffold}.ia.txt | grep $2_B | cut -d " " -f3)
    A_WPROP_INTRO=$(echo "scale=4; $A_LP_WIN/$TOT_WIN" | bc | sed 's/^\./0./')
    B_WPROP_INTRO=$(echo "scale=4; $B_LP_WIN/$TOT_WIN" | bc | sed 's/^\./0./')
    
    echo -e  "$INDIVIDUAL\t$SCAFFOLD\t$START\t$END\t$LENGTH\t$TOT_WIN\t$A_LP_WIN\t$B_LP_WIN\t$A_PCA_INTRO\t$B_PCA_INTRO\t$A_WPROP_INTRO\t$B_WPROP_INTRO" >> $2_Table1.txt
    
done

echo " - Table 1 for $2 in $1 Generated! -"

## Table 2

echo " - Generating Table 2 for $2 in $1... - "

for scaffold in $scaffolds
    do
    
    echo -ne " - ${scaffold} - \r"
	
	nscaf=$(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | cut -f1 | grep -o "Window" | wc -l)
		
	paste <(yes "$2" | head -$nscaf) <(yes "${scaffold}" | head -$nscaf) <(paste -d '.' <(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | cut -f1 | grep -o "Window") <(seq -f "%03g" 1 $(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | wc -l))) <(rev PCAdmix_${scaffold}/pcadmix_${scaffold}.vit.txt | cut -d " " -f3- | rev | grep $2_A | tr ' ' '\n' | grep -v "$2_A") <(rev PCAdmix_${scaffold}/pcadmix_${scaffold}.vit.txt | cut -d " " -f3- | rev | grep $2_B | tr ' ' '\n' | grep -v "$2_B") <(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | cut -f2- | cut -d " " -f1 | cut -d ":" -f2) <(awk -v FS=" " 'NF>20' PCAdmix_${scaffold}/pcadmix_${scaffold}.markers.txt | cut -f2- | cut -d " " -f20 | cut -d ":" -f2) > PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt
	
done

echo " - concatenating... -"

cat PCAdmix_*/*_$2_Table2.txt > $Admx_PATH/$1/$2_Table2.txt

echo " - Table 2 for $2 in $1 Generated! -"

## Table 3

echo " - Generating Table 3 for $2 in $1... - "

for scaffold in $scaffolds
    
    do
    	echo -ne " - Finding A allele introgressed segments in ${scaffold}... - \r"
    	
		awin=$(rev PCAdmix_${scaffold}/pcadmix_${scaffold}.vit.txt | cut -d " " -f3- | rev | grep $2_A | tr ' ' '\n' | grep -v "$2_A" | grep "1" | wc -l)
				
		for n in $(seq 1 $awin)
		
			do
				var=$(yes "Window...._1" | head -$n)
				lvar=$(echo $var)
	  
				cut -f3,4 PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt | tr '\t' '_' | tr '\n' ' ' | grep -o -e "Window...._0 $lvar Window...._0" | cut -d ' ' -f2- | rev | cut -d ' ' -f2- | rev | sed -e 's/_1//g' > PCAdmix_${scaffold}/$2_A_${n}

				cut -d ' ' -f1 PCAdmix_${scaffold}/$2_A_${n} > PCAdmix_${scaffold}/$2_A_${n}_starts

				rev PCAdmix_${scaffold}/$2_A_${n} | cut -d ' ' -f1 | rev > PCAdmix_${scaffold}/$2_A_${n}_ends
			
				cat PCAdmix_${scaffold}/$2_A_${n} | tr ' ' '-' | awk -vFS="-" '{print NF, $0}' | tr ' ' '\t' > PCAdmix_${scaffold}/$2_A_${n}_windows
				
				arows=$(cat PCAdmix_${scaffold}/$2_A_${n}_windows | wc -l)
				paste <(yes "$2" | head -$arows) <(yes "${scaffold}" | head -$arows) <(yes "A" | head -$arows) PCAdmix_${scaffold}/$2_A_${n}_windows <(grep -f PCAdmix_${scaffold}/$2_A_${n}_starts PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt | cut -f6) <(grep -f PCAdmix_${scaffold}/$2_A_${n}_ends PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt | cut -f7) > PCAdmix_${scaffold}/$2_A_${n}_Table3.txt

		done

		echo -ne " - Finding B allele introgressed segments in ${scaffold}... - \r"

		bwin=$(rev PCAdmix_${scaffold}/pcadmix_${scaffold}.vit.txt | cut -d " " -f3- | rev | grep $2_B | tr ' ' '\n' | grep -v "$2_B" | grep "1" | wc -l)

		for n in $(seq 1 $bwin)
		
			do
				var=$(yes "Window...._1" | head -$n)
				lvar=$(echo $var)

				cut -f3,5 PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt | tr '\t' '_' | tr '\n' ' ' | grep -o -e "Window...._0 $lvar Window...._0" | cut -d ' ' -f2- | rev | cut -d ' ' -f2- | rev | sed -e 's/_1//g' > PCAdmix_${scaffold}/$2_B_${n}

				cut -d ' ' -f1 PCAdmix_${scaffold}/$2_B_${n} > PCAdmix_${scaffold}/$2_B_${n}_starts

				rev PCAdmix_${scaffold}/$2_B_${n} | cut -d ' ' -f1 | rev > PCAdmix_${scaffold}/$2_B_${n}_ends

				cat PCAdmix_${scaffold}/$2_B_${n} | tr ' ' '-' | awk -vFS="-" '{print NF, $0}' | tr ' ' '\t' > PCAdmix_${scaffold}/$2_B_${n}_windows
		
				brows=$(cat PCAdmix_${scaffold}/$2_B_${n}_windows | wc -l)
				paste <(yes "$2" | head -$brows) <(yes "${scaffold}" | head -$brows) <(yes "B" | head -$brows) PCAdmix_${scaffold}/$2_B_${n}_windows <(grep -f PCAdmix_${scaffold}/$2_B_${n}_starts PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt | cut -f6) <(grep -f PCAdmix_${scaffold}/$2_B_${n}_ends PCAdmix_${scaffold}/pcadmix_${scaffold}_$2_Table2.txt | cut -f7) > PCAdmix_${scaffold}/$2_B_${n}_Table3.txt

		done
			
done

echo " - Introgressed segments identified! - "
echo " - Joining Table3s ... -"
cat PCAdmix_*/$2_A_*_Table3.txt > $2_A_Table3.txt
cat PCAdmix_*/$2_B_*_Table3.txt > $2_B_Table3.txt
echo -e "Individual\tScaffold\tAllele\tNWindows\tSegmentWindows\tStart\tEnd\tLength" > $Admx_PATH/$1/$2_Table3.txt
cat $2_A_Table3.txt $2_B_Table3.txt | sort -k3,3 -k2,2 -k6,6n | awk 'BEGIN { OFS = "\t" } { $8 = $7 - $6 } 1' >> $Admx_PATH/$1/$2_Table3.txt

echo " - Removing Intermediary files ... -"
rm PCAdmix_*/$2_*

echo " - Table 3 for $2 in $1 Generated! -"

```

## Loop for CR and KI Table of results generation

This loop can be run to extract the tables of all individuals in Particular Population (in this case CR)

```{r Tables, eval=FALSE, engine='bash'}
# First for PV individual (20SNPs analysis)
cd
./PCAdmix_TableMaker.sh PCAdmix_results h_ll_pv_0223

# Loop for all CR individuals in Carpathians results directory (20SNPs analysis)

crIDs=$(cat $Admx_PATH/cr_list_to_remove.txt)
for i in $crIDs
  do
    ./PCAdmix_TableMaker.sh PCAdmix_Carpathians $i
done

# Loop for all KI individuals in Kirov results directory (20SNPs analysis)

kiIDs=$(cat $Admx_PATH/ki_list_to_remove.txt)

for i in $kiIDs
  do
    ./PCAdmix_TableMaker.sh PCAdmix_Kirov $i
done

```

# 5. Table of Introgressed regions
Get table (from a VCF) with Information regarding the introgressed regions (similar to what done for Pelage and Epilepsy). Intersect a BED file with the introgressed regions coordinates with the VCF, then extract the table from that VCF.

## Generating BED file

We need to generate the BED file with the coordinates of the introgressed segments. Table3 of the PCAdmix results has these coordinates. By cutting the "useful" columns we can generate a BED file. Then we can intersect it with the VCF file containing the individual of interest

```{r Tables, eval=FALSE, engine='bash'}

grep "lp23" h_ll_pv_0223_Table3.txt | cut -f2,6,7 | sort -u > h_ll_pv_0223_introgressed.bed
grep "lp23" c_ll_cr_0205_Table3.txt | cut -f2,6,7 | sort -u > c_ll_cr_0205_introgressed.bed

bedtools intersect -header -a $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf -b h_ll_pv_0223_introgressed.bed > h_ll_pv_0223_introgressed.vcf
bedtools intersect -header -a $Data_PATH/ll_west_perpop_filtered2_1mbpnoxyscaffolds_intersected.lr_ann.vcf -b c_ll_cr_0205_introgressed.bed > c_ll_cr_0205_introgressed.vcf

grep "CDS" h_ll_pv_0223_introgressed.vcf | cut -d'|' -f4 > h_ll_pv_0223_introgressed.genelist
grep "CDS" c_ll_cr_0205_introgressed.vcf | cut -d'|' -f4 > c_ll_cr_0205_introgressed.genelist

grep -f h_ll_pv_0223_introgressed.genelist ~/Data_b/LYPA23C.all.fix.gff3 > h_ll_pv_0223_introgressed.genelist.gff3
grep -f c_ll_cr_0205_introgressed.genelist ~/Data_b/LYPA23C.all.fix.gff3 > c_ll_cr_0205_introgressed.genelist.gff3

```
