---
title: "PCAdmix"
author: "Enrico"
date: "5 December 2018"
output: html_document
---
# 0. Path Definition

Define the paths to different locations for easier calling during script

```{r Define paths, eval=FALSE, engine='bash'}

ssh -X ebazzicalupo@genomics-a.ebd.csic.es

Admx_PATH=/home/ebazzicalupo/Admixture # path to the project directory
Data_PATH=/home/ebazzicalupo/Data # path to data directory
ShapeIt=/home/ebazzicalupo/Admixture/ShapeIt/shapeit # path to ShapeIt software
Lenght_PATH=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/Length_scaffolds_lp23 # path to scaffold length file
SexChr_PATH=/home/GRUPOS/grupolince/sexual_chr # path to directory with lists of scaffolds with XY chromosomes represented
BCF=/opt/bcftools-1.6/bcftools # path to BCFtools software
REF=/home/GRUPOS/grupolince/reference_genomes/lynx_pardinus_genome/lp23.fa # path to reference genome
GATK=/opt/GATK-3.7/GenomeAnalysisTK.jar # GATK software path
VCF2BEA=/home/ebazzicalupo/Admixture/VCF2Beagle/vcf2beagle.jar # vcf2beagle software path
PCAdmix=/home/ebazzicalupo/Admixture/PCAdmix/PCAdmix3_linux # PCAdmix software path

```

# 3. PCAdmix

Run the PCAdmix software for each phased scaffold. Different populations have been tested with different options. Lynx lynx from PV, CR population and KI populations were analyzed. The analysis was always run with Lynx pardinus from SM and Lynx lynx from the EAST as ancestral populations. Options: -w 20 for 20SNPs window size; -lab for labels, but not really working.

## Version with PV population as admixed

```{r PCAdmix, eval=FALSE, engine='bash'}

# loop for all scaffolds
# WITHOUT PUTTING A MAP IT WILL USE DEFAULT WINDOW SIZE OF 20 SNPs (-wMb is useless without a map)

screen -S PCAdmix.log
script PCAdmix.log

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_results/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_pv_${scaffold}_Phase/ll_pv_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_results/PCAdmix_${scaffold}/pcadmix_${scaffold} -lab LYNX PARDINUS Admixed -wMb 0.1
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_results/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_results/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_results/PCAdmix_lp23.s*/*vit.txt > ~/home/allscaffolds.txt

grep "1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1" ~/home/allscaffolds.txt | wc -l

```


## Version with Carpathians population as Admixed (for results comparison)

First we will need extract a phased vcf with only the CR population and convert it to beagle format. Then we can run PCAdmix with CR as admixed (same options and ancestral populations as PV).

```{r PCAdmix, eval=FALSE, engine='bash'}

## extract list of CR individuals

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'cr' > $Admx_PATH/cr_list_to_remove.txt
cat $Admx_PATH/cr_list_to_remove.txt

cd $Admx_PATH
screen -S CRpop.log
script CRpop.log

## for each scaffold: create a new directory just for the CR population, extract CR population information from West population phased vcf using the function view of BCFtools (zipping and unzipping and indexing done in order to make the software work)

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo ${scaffold}
    rm $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz.tbi
    bgzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf
    tabix -p vcf $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    rm -r $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase
    mkdir $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase
    
    $BCF view -Ov -S $Admx_PATH/cr_list_to_remove.txt $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz > $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.vcf
    gunzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    echo " DONE "

done

## convert the CR-only phased vcf to beagle

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo  " converting to beagle ${scaffold} in cr "
    cat $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.vcf | java -jar $VCF2BEA N $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased
    gunzip $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.bgl.gz
    echo " DONE "
done

## run 20SNPs window PCAdmix

screen -S CR_PCAdmix.log
script CR_PCAdmix.log
mkdir $Admx_PATH/PCAdmix_Carpathians

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_Carpathians/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_cr_${scaffold}_Phase/ll_cr_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_Carpathians/PCAdmix_${scaffold}/pcadmix_${scaffold} -lab LYNX PARDINUS Admixed -w 20
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_Carpathians/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_Carpathians/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_Carpathians/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/Carpathians.txt

```

## Version with Kirov population as Admixed (for results comparison)

First we will need extract a phased vcf with only the KI population and convert it to beagle format. Then we can run PCAdmix with KI as admixed (same options and ancestral populations as PV).


```{r PCAdmix, eval=FALSE, engine='bash'}

## extract list of KI individuals

$BCF query -l $Data_PATH/ll_perspecies_filtered2_1mbpnoxyscaffolds.lr_ann.vcf | grep 'ki' > $Admx_PATH/ki_list_to_remove.txt
cat $Admx_PATH/ki_list_to_remove.txt

cd $Admx_PATH
screen -S KIpop.log
script KIpop.log

## for each scaffold: create a new directory just for the KI population, extract KI population information from West population phased vcf using the function view of BCFtools (zipping and unzipping and indexing done in order to make the software work)

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo ${scaffold}
    rm $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz.tbi
    bgzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf
    tabix -p vcf $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    rm -r $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase
    mkdir $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase
    
    $BCF view -Ov -S $Admx_PATH/ki_list_to_remove.txt $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz > $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.vcf
    gunzip $Admx_PATH/Phasing/ll_west_${scaffold}_Phase/ll_west_${scaffold}.phased.vcf.gz
    
    echo " DONE "

done

## convert the KI-only phased vcf to beagle

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo  " converting to beagle ${scaffold} in ki "
    cat $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.vcf | java -jar $VCF2BEA N $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased
    gunzip $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.bgl.gz
    echo " DONE "
done

## run 20SNPs window PCAdmix

screen -S ki_PCAdmix.log
script ki_PCAdmix.log
mkdir $Admx_PATH/PCAdmix_Kirov

scaffolds=$(cat $Admx_PATH/lp23_scaffolds_tobephased_noerror)
for scaffold in $scaffolds
    do
    echo " starting PCAdmix for ${scaffold} "
    mkdir $Admx_PATH/PCAdmix_Kirov/PCAdmix_${scaffold}
    
    $PCAdmix -anc $Admx_PATH/Phasing/ll_east_${scaffold}_Phase/ll_east_${scaffold}.phased.bgl $Admx_PATH/Phasing/lp_sm_${scaffold}_Phase/lp_sm_${scaffold}.phased.bgl -adm $Admx_PATH/Phasing/ll_ki_${scaffold}_Phase/ll_ki_${scaffold}.phased.bgl -o $Admx_PATH/PCAdmix_Kirov/PCAdmix_${scaffold}/pcadmix_${scaffold} -lab LYNX PARDINUS Admixed -w 20
    
    echo " DONE "
done

cat $Admx_PATH/PCAdmix_Kirov/PCAdmix_lp23.s*/*vit.txt | less -S
cat $Admx_PATH/PCAdmix_Kirov/PCAdmix_lp23.s*/*ia.txt | less -S

cat ~/home/PCAdmix_Kirov/PCAdmix_lp23.s*/*vit.txt | cut -d' ' -f2- > ~/home/Kirov.txt

```